### PR TITLE
feat: Enable Vector database and retrieve_online_documents API

### PR DESCRIPTION
**Pull request by HaoXuAI**
**What this PR does / why we need it**:
RFC: https://docs.google.com/document/d/18IWzLEA9i2lDWnbfbwXnMCg3StlqaLVI-uRpQjr_Vos/edit#heading=h.9gaqqtox9jg6

**Which issue(s) this PR fixes**:

Fixes #

https://github.com/feast-dev/feast/issues/3965

### CODE REVIEW

**Review by franciscojavierarceo — CHANGES_REQUESTED** [review]
Overall this looks great, I left some comments in the doc and think the only thing that's worth doing is us aligning on some of the naming conventions to be more similar to the industry and to what Feast is already doing.

Great work and can't wait to get this out! 🚀

File: sdk/python/feast/feature_store.py [hunk 1]
```diff
@@ -1690,6 +1690,67 @@ def _get_online_features(
...
 
+    @log_exceptions_and_usage
+    def retrieve_online_documents(
+        self,
+        feature: str,
+        query: Union[str, List[float]],
+        top_k: int,
+    ) -> OnlineResponse:
+        """
+        Retrieves the top k closest document features.
```
Comment (franciscojavierarceo) [inline sdk/python/feast/feature_store.py:1701]:
```suggestion
        Retrieves the top k closest document features. Note, embeddings are a subset of features.
```

File: sdk/python/feast/feature_store.py [hunk 2]
```diff
@@ -1690,6 +1690,67 @@ def _get_online_features(
...
+
+        Args:
+            feature: The list of document features that should be retrieved from the online document store. These features can be
+                specified either as a list of string document feature references or as a feature service. String feature
+                references must have format "feature_view:feature", e.g, "document_fv:document_embedding_feature".
```
Comment (franciscojavierarceo) [inline sdk/python/feast/feature_store.py:1706]:
```suggestion
                references must have format "feature_view:feature", e.g, "document_fv:document_embeddings".
```

File: sdk/python/feast/feature_store.py [hunk 3]
```diff
@@ -1690,6 +1690,67 @@ def _get_online_features(
         )
         return OnlineResponse(online_features_response)
 
+    @log_exceptions_and_usage
+    def retrieve_online_documents(
```
Comment (franciscojavierarceo) [inline sdk/python/feast/feature_store.py:1694]: There's probably something to be said about having a configurable distance metric to let the user choose which way to get the `top_k`

File: sdk/python/feast/infra/online_stores/contrib/postgres.py [hunk 4]
```diff
@@ -251,6 +260,49 @@ def teardown(
...
+            requested_feature: The requested feature as the column to search
+            embedding: The query embedding to search for
+            top_k: The number of items to return
+        Returns:
+            List of tuples containing the event timestamp and the document feature
+
+        """
+
+        # Convert the embedding to a string to be used in postgres vector search
+        query_embedding_str = f"'[{','.join(str(el) for el in embedding)}]'"
```
Comment (franciscojavierarceo) [inline sdk/python/feast/infra/online_stores/contrib/postgres.py:285]: Is this the best serialization we can do? This feels pretty brittle but I get it.

File: sdk/python/feast/infra/passthrough_provider.py [hunk 5]
```diff
@@ -47,6 +47,7 @@ def __init__(self, config: RepoConfig):
         self.repo_config = config
         self._offline_store = None
         self._online_store = None
+        self._document_store = None
```
Comment (tokoko) [inline sdk/python/feast/infra/passthrough_provider.py:50]: This is no longer necessary

File: sdk/python/feast/infra/online_stores/contrib/postgres.py [hunk 6]
```diff
@@ -251,6 +260,49 @@ def teardown(
...
             raise
 
+    def retrieve_online_documents(
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        requested_feature: str,
+        embedding: List[float],
+        top_k: int,
+    ) -> List[Tuple[Optional[datetime], Optional[Dict[str, ValueProto]]]]:
```
Comment (tokoko) [inline sdk/python/feast/infra/online_stores/contrib/postgres.py:270]: Do we need `Optional[Dict[str, ValueProto]]`? If only a single feature can be searched, wouldn't `Optional[ValueProto]` be sufficient?

File: sdk/python/tests/integration/online_store/test_universal_online.py [hunk 7]
```diff
@@ -785,3 +785,26 @@ def assert_feature_service_entity_mapping_correctness(
                 entity_rows=entity_rows,
                 full_feature_names=full_feature_names,
             )
+
+
+@pytest.mark.integration
+@pytest.mark.universal_online_stores(only=["postgres"])
+def test_retrieve_online_documents(
```
Comment (franciscojavierarceo) [inline sdk/python/tests/integration/online_store/test_universal_online.py:792]: Will you be outputting the cosine similarity as well? That would be useful possibly for debugging. Would be good to be able to test that the engine computes it...maybe not doable though.

Comment (HaoXuAI) [reply sdk/python/tests/integration/online_store/test_universal_online.py:792]: Should be possible. Somehow just the integration test doesn't startup the Postgres container. And I'm debugging it.

**Review by franciscojavierarceo — APPROVED** [review]
lgtm

**Comment by HaoXuAI** [comment]
> lgtm

thanks!

**Comment by HaoXuAI** [comment]
@franciscojavierarceo @tokoko fyi this is the result of the API. there are still some TODOs I've added into the code directly. Will add to the document as well. Also feels like having a layer to abstract the `vector` online store might help to control the input and output flow.  Anyway we have a working solution now.

Screenshot 2024-04-14 at 3 24 34 PM
Screenshot 2024-04-14 at 3 24 30 PM

File: sdk/python/feast/infra/passthrough_provider.py (hunk 5 above)
Comment (boumelhaa) [reply sdk/python/feast/infra/passthrough_provider.py]: Thanks for addressing the issue. One question, though: what if we want to continue using Redis or any other online store for usual features, and use PG vector solely for embedding and search? Do we have the option to use the online store and the document store in the feature_store.yaml, both?

Comment (franciscojavierarceo) [reply sdk/python/feast/infra/passthrough_provider.py]: That could get complicated but agreed it'd be good to add. I could imagine a Redis + another DB layer would be an obvious one.

Comment (boumelhaa) [reply sdk/python/feast/infra/passthrough_provider.py]: Thank you again, guys, for these amazing features. Yes, having multiple online stores will make it easier to use the right database layer for the appropriate use case! 🙌

File: Makefile [hunk 8]
```diff
@@ -200,7 +200,7 @@ test-python-universal-postgres-offline:
 test-python-universal-postgres-online:
 	PYTHONPATH='.' \
 		FULL_REPO_CONFIGS_MODULE=sdk.python.feast.infra.online_stores.contrib.postgres_repo_configuration \
-		PYTEST_PLUGINS=sdk.python.feast.infra.offline_stores.contrib.postgres_offline_store.tests \
```
Comment (franciscojavierarceo) [inline Makefile:203]: Looking at this now, was this the right choice?

**Pull request merged**